### PR TITLE
[style] advance C++ language standard to C++11

### LINKED ISSUE #3880: Move language standard baseline to C++11?
The current language standard baseline is C++03 (see [STYLE_GUIDE.md](https://github.com/openthread/openthread/blob/master/STYLE_GUIDE.md#standards)).

Thoughts on whether we should move to supporting C++11?

One use case was raised in #3877.

### PR DESCRIPTION
**Pull request by jwhui**
This PR makes the following changes:
- Update `-std` to `c++11`.
- Add `-pedantic-errors` to avoid use of gcc extensions.
- Update STYLE_GUIDE.md.
- Remove use of zero-length args for variadic macro (a gcc extension).

Resolves #3880.

### CODE REVIEW

File: src/core/api/ip6_api.cpp [hunk 1]
```diff
@@ -195,12 +195,8 @@ otMessage *otIp6NewMessageFromBuffer(otInstance *             aInstance,
                                      const otMessageSettings *aSettings)
 {
     Instance &instance = *static_cast<Instance *>(aInstance);
-    Message * message;
-
-    VerifyOrExit((message = instance.Get<Ip6::Ip6>().NewMessage(aData, aDataLength, aSettings)) != NULL);
 
-exit:
-    return message;
+    return instance.Get<Ip6::Ip6>().NewMessage(aData, aDataLength, aSettings);
```
Comment (abtink) [inline src/core/api/ip6_api.cpp]: :+1:

**Comment by bukepo** [comment]
Also switch to the standard static_assert?

File: configure.ac [hunk 2]
```diff
@@ -233,7 +233,7 @@ AC_PATH_PROG(CMP, cmp)
 #
 
 PROSPECTIVE_CFLAGS="-Wall -Wextra -Wshadow -Werror -std=c99 -pedantic-errors"
-PROSPECTIVE_CXXFLAGS="-Wall -Wextra -Wshadow -Werror -std=gnu++98 -Wno-c++14-compat -fno-exceptions"
+PROSPECTIVE_CXXFLAGS="-Wall -Wextra -Wshadow -Werror -std=c++11 -pedantic-errors -fno-exceptions"
```
Comment (KaSroka) [inline configure.ac]: We should define c++11 for Android build as well probably.

Comment (KaSroka) [reply configure.ac]: And pedantic-errors as well.

**Comment by LuDuda** [comment]
Thanks for this initiative. I would like to run IAR, Keil and SES compilers on this PR before merge if possible. I'm ooo this week but I will try to do that on Monday if that is okay.

File: configure.ac (hunk 2 above)
Comment (jwhui) [reply configure.ac]: I updated `LOCAL_CPPFLAGS` in `Android.mk`.

**Comment by LuDuda** [comment]
So far the build fails, but I need to spend more time to update project generation templates (adding some compiler flags etc.). I will inform you about the progress.

**Comment by jwhui** [comment]
@LuDuda, thanks for the update and helping with this PR!

**Comment by LuDuda** [comment]
Hello @jwhui,

I have verified that C++11 is supported by most of the compilers such as GCC, SES or Keil. During the investigation, I found that IAR does not handle it correctly. This is due to the fact that Nordic supports version 7.80 of the toolchain in the nRF5 SDK, which is also widely used by our customers (i.e. customers of OpenThread).
The support for C++11/14 has been introduced in version 8.10: https://www.iar.com/support/resources/articles/exploring-c11-and-c14/

After some internal discussions, we don't plan to update the version of the toolchain and so Nordic's recommendation is to stay with the current version of the C++.

In case you decide to move this PR forward, I quickly verified that it would be possible to deliver OT libraries (built under IAR 8.10) and use them with older IAR version, although any configuration changes etc. would require some effort from the customers.

**Comment by jwhui** [comment]
Based on feedback above, will not pursue C++11 for now.

**Comment by bukepo** [comment]
@LuDuda do you have any plan update to a newer version supporting c++11 now?

**Comment by LuDuda** [comment]
Hello, @bukepo I'll be back with some answers as soon as possible.

**Comment by bukepo** [comment]
@LuDuda any update for this issue?

**Comment by LuDuda** [comment]
Hello @bukepo @jwhui,

Nordic is okay now to move to C++11 standard in OpenThread core.
We have recently released our SDK which required C++03, from now on this should not be the issue anymore.

**Comment by bukepo** [comment]
> Hello @bukepo @jwhui,
> 
> 
> 
> Nordic is okay now to move to C++11 standard in OpenThread core.
> 
> We have recently released our SDK which required C++03, from now on this should not be the issue anymore.

Great news!

**Pull request closed without merging**